---
title: "MPXV_PRJEB56841"
author: "Lyskova_Alisa"
date: "2024-07-31"
output: html_document
---
```{r}
library(tidyverse)
library(showtext)
library(showtextdb)
showtext_auto()
font_add_google("Roboto Condensed")
```
```{r}
pink <- '#C6878F'
brown <- '#E07A5F'
blue <- '#3D405B'
green1 <- '#5F797B'
green2 <- '#81B29A'
yellow <- '#F2CC8F'
```

```{r setup, include=FALSE}
file_in <- c("PRJEB56841/clair3_vcf/summary_table_vcf.csv")
df <- read.csv(file_in, sep = "\t", header = TRUE)
df2 <- df %>% filter(file_type == "fastq.gz")

p <- df2 %>% mutate(allele = fct_relevel(allele, 'TCT', 'TCA', 'TCC', 'TCG', '*')) %>%
  mutate(time = fct_relevel(time, '1h', '2h', '4h', '6h', '12h', '24h')) %>%
  ggplot(aes(x=time, y=Nreads, fill=allele)) +
  geom_bar(position="dodge", stat="identity") +
  facet_wrap(~group) +
  theme_bw() +
    labs(title= 'Number of reads for mutation C>T and G>A in samples by time compared to all other types of mutations',
      x = 'Time',
      y = 'Number of reads') +
    theme(text=element_text(size=32, family="Roboto Condensed"),
        plot.title = element_text(hjust = 0.5, size = 36),
        axis.text=element_text(size=32)) +
    scale_fill_manual(values=c(brown, blue, yellow, green2, pink))

ggsave('PRJEB56841/clair3_vcf/summary_vcf_Nreads_fastq.gz.png', plot = p, height=3, width=10, dpi=300)

p
```


```{r setup, include=FALSE}
file_in <- c("PRJEB56841/clair3_vcf/summary_table_vcf.csv")
df <- read.csv(file_in, sep = "\t", header = TRUE)
df2 <- df %>% filter(file_type == "bam")

p <- df2 %>% mutate(allele = fct_relevel(allele, 'TCT', 'TCA', 'TCC', 'TCG', '*')) %>%
  mutate(time = fct_relevel(time, '1h', '2h', '4h', '6h', '12h', '24h')) %>%
  ggplot(aes(x=time, y=freq, fill=allele)) +
  geom_bar(position="dodge", stat="identity") +
  facet_wrap(~group) +
  theme_bw() +
    labs(title= 'Allele frequency for mutation C>T and G>A in samples by time compared to all other types of mutations',
      x = 'Time',
      y = 'Allele frequency') +
    theme(text=element_text(size=32, family="Roboto Condensed"),
        plot.title = element_text(hjust = 0.5, size = 36),
        axis.text=element_text(size=32)) +
    scale_fill_manual(values=c(brown, blue, yellow, green2, pink))

ggsave('PRJEB56841/clair3_vcf/summary_vcf_freq_bam.png', plot = p, height=3, width=10, dpi=300)

p
```


```{r setup, include=FALSE}
file_in <- c("PRJEB56841/clair3_vcf/summary_table_vcf_dRNA.csv")
df <- read.csv(file_in, sep = "\t", header = TRUE)
df2 <- df %>% filter(file_type == "fastq.gz")

pal <- c(brown, blue, yellow, green2, pink)

p <- df2 %>% mutate(allele = fct_relevel(allele, 'TCT', 'TCA', 'TCC', 'TCG', '*')) %>%
  ggplot(aes(x=allele, y=freq)) +
  geom_bar(position="dodge", stat="identity", fill=pal) +
  theme_bw() +
    labs(title= 'Allele frequency for C>T and G>A in sample ERR10963128 compared to all other types of mutations',
      x = 'Allele',
      y = 'Allele frequency') +
    theme(text=element_text(size=24, family="Roboto Condensed"),
        plot.title = element_text(hjust = 0.5, size = 24),
        axis.text=element_text(size=24))

ggsave('PRJEB56841/clair3_vcf/vcf_freq_dRNA_fastq.gz.png', plot = p, height=3, width=5, dpi=300)

p
```

